### PR TITLE
DEX-1678 order v4 feature

### PR DESCRIPTION
**Pull request by stavegan**

### CODE REVIEW

File: waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransaction.scala [hunk 1]
```diff
@@ -134,4 +137,45 @@ object ExchangeTransaction {
           ExchangeTransactionV3.parseBytes(bytes).map(_._1)
       }
 
+  //converts asset_decimal price to fixed_decimal
+  def convertPrice(price: Long, amountAssetDecimals: Int, priceAssetDecimals: Int): Either[GenericError, Long] =
```
Comment (JennaWestenra) [inline waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransaction.scala:141]: Maybe name it convertPriceToFixedDecimals?

File: dex/src/main/resources/application.conf [hunk 2]
```diff
@@ -192,6 +192,9 @@ waves.dex {
   # All transactions issued from this offset will have a different timestamps
   exchange-tx-ts-start-offset = 9223372036854775808
 
+  # Offset, starting from which orders of the fourth version will be issued
```
[the remark this answers is not in the document]
Comment (stavegan) [reply dex/src/main/resources/application.conf:195]: Synonyms: they'll be accepted to be issued.

File: waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransaction.scala (hunk 1 above)
Comment (stavegan) [reply waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransaction.scala:141]: Comment & function arguments look quite expressive in my opinion. Anyway, it's just rollback of one of the previous versions and not worth a 45 minute flapping build.

File: dex/src/test/scala/com/wavesplatform/dex/MatcherSpecBase.scala [hunk 3]
```diff
@@ -550,27 +550,29 @@ trait MatcherSpecBase
...
+        sellOrder = sellOrder,
+        amount = sellOrder.amount,
+        price = sellOrder.price,
+        buyMatcherFee = buyOrder.matcherFee,
+        sellMatcherFee = sellOrder.matcherFee,
+        fee = 300000L,
+        timestamp = nowTs,
+        proofs = Proofs.empty
+      )
+      .asInstanceOf[ExchangeTransactionResult[ExchangeTransaction]]
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/MatcherSpecBase.scala:575]: asInstanceOf can be removed

File: dex/src/test/scala/com/wavesplatform/dex/api/ws/ActorsWebSocketInteractionsSpecification.scala [hunk 4]
```diff
@@ -147,7 +147,9 @@ class ActorsWebSocketInteractionsSpecification
           fee = 300000L,
           timestamp = nowTs,
           proofs = Proofs.empty
-        ).copy(error = ValidationError.GenericError("Some error").some)
+        )
+        .copy(error = ValidationError.GenericError("Some error").some)
+        .asInstanceOf[ExchangeTransactionResult[ExchangeTransaction]]
```
Comment (alexander-branevskiy) [inline dex/src/test/scala/com/wavesplatform/dex/api/ws/ActorsWebSocketInteractionsSpecification.scala:152]: asInstanceOf can be removed

File: waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransactionResult.scala [hunk 5]
```diff
@@ -5,15 +5,15 @@ import cats.syntax.option._
 import com.wavesplatform.dex.domain.error.ValidationError
 
 // contains transaction and possible validation error
-final case class ExchangeTransactionResult[A <: ExchangeTransaction](transaction: A, error: Option[ValidationError] = None) {
+final case class ExchangeTransactionResult[+A <: ExchangeTransaction](transaction: A, error: Option[ValidationError] = None) {
```
Comment (alexander-branevskiy) [inline waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransactionResult.scala:8]: nice 👍

File: waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransactionV2.scala [hunk 6]
```diff
@@ -40,6 +43,64 @@ case class ExchangeTransactionV2(
 
 object ExchangeTransactionV2 extends ExchangeTransactionParser[ExchangeTransactionV2] {
 
+  def create(
```
Comment (alexander-branevskiy) [inline waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransactionV2.scala:46]: mkSigned

File: waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransactionV2.scala [hunk 7]
```diff
@@ -40,6 +43,64 @@ case class ExchangeTransactionV2(
...
+      buyMatcherFee,
+      sellMatcherFee,
+      fee,
+      timestamp,
+      proofs = Proofs.empty
+    ).map { unverified =>
+      unverified.copy(proofs = Proofs(List(ByteStr(crypto.sign(matcher, unverified.bodyBytes())))))
+    }
+
+  def create(
```
Comment (alexander-branevskiy) [inline waves-integration/src/main/scala/com/wavesplatform/dex/domain/transaction/ExchangeTransactionV2.scala:75]: mk

**Review by alexander-branevskiy — CHANGES_REQUESTED** [review]

**Review by alexander-branevskiy — APPROVED** [review]

**Pull request merged**